File: iosApp/iosApp/Presentation/Main/MainViewModel.swift
//
//  MainViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/07.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class MainViewModel: ObservableObject {
    private let getMainNavigationUseCase: GetMainNavigationUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let restoreSessionUseCase: RestoreSessionUseCase

    @Published private(set) var uiState = MainUiState.empty

    let event = PassthroughSubject<MainEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func refreshNavigation(preferredRoute: String?) {
        tasks[.refreshNavigation]?.cancel()
        tasks[.refreshNavigation] = Task {
            do {
                let result = try await getMainNavigationUseCase.invoke(preferredRoute: preferredRoute)
                guard !Task.isCancelled else { return }

                if let success = result as? AppResultSuccess<AnyObject>,
                   let state = success.data as? MainNavigationState {
                    if state.currentUser?.signupCompleted == false {
                        event.send(.navigateToSignUp)
                    }
                    uiState = MainUiState(
                        currentUser: state.currentUser,
                        tabs: state.tabs,
                        selectedTab: state.selectedTab,
                        thirdTab: state.thirdTab
                    )
                } else if let failure = result as? AppResultFailure {
                    event.send(.showError(message: "\(failure.error)"))
                }
            } catch {
                guard !Task.isCancelled else { return }
                event.send(.showError(message: error.localizedDescription))
            }
        }
    }

    private func selectTab(_ route: String) {
        guard uiState.tabs.contains(where: { $0.route == route }) || route == MainNavigationTab.community.route else { return }
        uiState.selectedTab = route
    }

    private func observeSession() {
        tasks[.observeSession]?.cancel()
        tasks[.observeSession] = Task {
            do {
                for try await newUser in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    guard !Task.isCancelled else { return }
                    let currentUser = self.uiState.currentUser
                    let roleChanged = newUser?.role != currentUser?.role
                    let loginStateChanged = (newUser == nil) != (currentUser == nil)

                    if roleChanged || loginStateChanged {
                        // Only rebuild navigation when login state or role changes to avoid
                        // spurious selectedTab resets caused by Firestore/auth re-emissions.
                        self.refreshNavigation(preferredRoute: self.uiState.selectedTab)
                    } else {
                        // Same user/role — just sync the user object without touching the tab.
                        self.uiState.currentUser = newUser
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func restoreSession() {
        Task {
            do {
                _ = try await restoreSessionUseCase.invoke()
                refreshNavigation(preferredRoute: uiState.selectedTab)
            } catch {
                refreshNavigation(preferredRoute: uiState.selectedTab)
            }
        }
    }

    func onAction(_ action: MainAction) {
        switch action {
        case .enter(let preferredRoute):
            refreshNavigation(preferredRoute: preferredRoute)
        case .refreshNavigation(let preferredRoute):
            refreshNavigation(preferredRoute: preferredRoute)
        case .selectTab(let route):
            selectTab(route)
        }
    }

    init(
        getMainNavigationUseCase: GetMainNavigationUseCase = KoinInitializerKt.resolveGetMainNavigationUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        restoreSessionUseCase: RestoreSessionUseCase = KoinInitializerKt.resolveRestoreSessionUseCase()
    ) {
        self.getMainNavigationUseCase = getMainNavigationUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.restoreSessionUseCase = restoreSessionUseCase

        observeSession()
        restoreSession()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case observeSession
        case refreshNavigation
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/MainViewModel.kt
package com.hhp227.concafe.presentation.main

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.MainNavigationTab
import com.hhp227.concafe.domain.usecase.CheckAppUpdateUseCase
import com.hhp227.concafe.domain.usecase.GetMainNavigationUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import com.hhp227.concafe.domain.usecase.RestoreSessionUseCase

class MainViewModel(
    private val checkAppUpdateUseCase: CheckAppUpdateUseCase,
    private val getMainNavigationUseCase: GetMainNavigationUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val restoreSessionUseCase: RestoreSessionUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(MainUiState.empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<MainEvent>()
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<JobKey, Job>()

    private fun refreshNavigation(preferredRoute: String? = null) {
        jobs[JobKey.REFRESH_NAVIGATION]?.cancel()
        jobs[JobKey.REFRESH_NAVIGATION] = viewModelScope.launch {
            when (val result = getMainNavigationUseCase.invoke(preferredRoute)) {
                is AppResult.Success -> {
                    if (result.data.currentUser?.signupCompleted == false) {
                        _event.emit(MainEvent.NavigateToSignUp)
                    }
                    _uiState.update {
                        it.copy(
                            currentUser = result.data.currentUser,
                            tabs = result.data.tabs,
                            selectedTab = result.data.selectedTab,
                            thirdTab = result.data.thirdTab
                        )
                    }
                }
                is AppResult.Failure -> {
                    _event.emit(MainEvent.ShowError(result.error.toString()))
                }
            }
        }
    }

    private fun selectTab(route: String) {
        _uiState.update { state ->
            if (state.tabs.any { it.route == route } || route == MainNavigationTab.COMMUNITY.route) {
                state.copy(selectedTab = route)
            } else {
                state
            }
        }
    }

    private fun observeSession() {
        jobs[JobKey.OBSERVE_SESSION]?.cancel()
        jobs[JobKey.OBSERVE_SESSION] = viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest { newUser ->
                val currentState = _uiState.value
                val roleChanged = newUser?.role != currentState.currentUser?.role
                val loginStateChanged = (newUser == null) != (currentState.currentUser == null)

                if (roleChanged || loginStateChanged) {
                    // Only rebuild navigation when login state or role changes to avoid
                    // spurious selectedTab resets caused by Firestore/auth re-emissions.
                    refreshNavigation(currentState.selectedTab)
                } else {
                    // Same user/role — just sync the user object without touching the tab.
                    _uiState.update { it.copy(currentUser = newUser) }
                }
            }
        }
    }

    private fun restoreSession() {
        viewModelScope.launch {
            when (restoreSessionUseCase.invoke()) {
                is AppResult.Success -> refreshNavigation(_uiState.value.selectedTab)
                is AppResult.Failure -> refreshNavigation(_uiState.value.selectedTab)
            }
        }
    }

    private fun checkAppUpdate(storePlatform: String, storeId: String, currentVersion: String) {
        jobs[JobKey.CHECK_APP_UPDATE]?.cancel()
        jobs[JobKey.CHECK_APP_UPDATE] = viewModelScope.launch {
            when (
                val result = checkAppUpdateUseCase.invoke(
                    storePlatform = storePlatform,
                    storeId = storeId,
                    currentVersion = currentVersion
                )
            ) {
                is AppResult.Success -> result.data?.let { _event.emit(MainEvent.ShowAppUpdate(it)) }
                is AppResult.Failure -> Unit
            }
        }
    }

    fun onAction(action: MainAction) {
        when (action) {
            is MainAction.CheckAppUpdate -> checkAppUpdate(
                storePlatform = action.storePlatform,
                storeId = action.storeId,
                currentVersion = action.currentVersion
            )
            is MainAction.Enter -> refreshNavigation(action.preferredRoute)
            is MainAction.RefreshNavigation -> refreshNavigation(action.preferredRoute)
            is MainAction.SelectTab -> selectTab(action.route)
        }
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    init {
        observeSession()
        restoreSession()
    }

    private enum class JobKey {
        CHECK_APP_UPDATE,
        OBSERVE_SESSION,
        REFRESH_NAVIGATION
    }
}
